### PR TITLE
Add Distributed VectorCache and use it for intermediate vector in and adapt `apply_uses_initial_guess` in Schwarz 

### PR DESCRIPTION
**Pull request by yhmtsai**
This adds a cache/workspace for the intermediate vector in advanced_apply of Schwarz.
Also, it makes apply_uses_initial_guess rely on the `local_solver->apply_uses_initial_guess()`.
We might need to revisit this function after adding refining step in Schwarz.

### CODE REVIEW

File: core/distributed/preconditioner/schwarz.cpp [hunk 1]
```diff
@@ -84,15 +91,27 @@ void Schwarz<ValueType, LocalIndexType, GlobalIndexType>::apply_impl(
...
-            dense_x->add_scaled(dense_alpha, x_clone.get());
+            dense_x->add_scaled(dense_alpha, cache_.intermediate.get());
         },
         alpha, b, beta, x);
 }
 
 
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+template <typename VectorType>
+void Schwarz<ValueType, LocalIndexType, GlobalIndexType>::set_cache_to(
```
Comment (MarcelKoch) [inline core/distributed/preconditioner/schwarz.cpp:105]: You only need this function once, right? Then I would prefer to just add this directly, it's not that much IMO.

File: include/ginkgo/core/distributed/preconditioner/schwarz.hpp [hunk 2]
```diff
@@ -150,6 +169,26 @@ class Schwarz
...
+     * Manages a vector as a cache, so there is no need to allocate one every
+     * time an intermediate vector is required.
+     * Copying an instance will only yield an empty object since copying the
+     * cached vector would not make sense.
+     *
+     * @internal  The struct is present so the whole class can be copyable
+     *            (could also be done with writing `operator=` and copy
+     *            constructor of the enclosing class by hand)
+     */
+    mutable struct cache_struct {
```
Comment (MarcelKoch) [inline include/ginkgo/core/distributed/preconditioner/schwarz.hpp:183]: I think this should be a `DenseCache` instead.

**Review by MarcelKoch — CHANGES_REQUESTED** [review]

**Review by pratikvn — APPROVED** [review]
I agree with the Marcel that it can be a `DenseCache` instead of `cache_struct`. But otherwise looks good to me. I approve it as Marcel has already requested that change.

File: include/ginkgo/core/distributed/preconditioner/schwarz.hpp (hunk 2 above)
Comment (yhmtsai) [reply include/ginkgo/core/distributed/preconditioner/schwarz.hpp:183]: I think it does not support distributed::Vector.

Comment (MarcelKoch) [reply include/ginkgo/core/distributed/preconditioner/schwarz.hpp:183]: you're right. Perhaps it would make sense to add `VectorCache`, instead of using the cache construct.

File: core/base/vector_cache.cpp [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+void VectorCache<ValueType>::init(std::shared_ptr<const Executor> exec,
+                                  gko::experimental::mpi::communicator comm,
+                                  dim<2> global_size, dim<2> local_size) const
+{
+    if (!vec || vec->get_size() != global_size || vec->get_executor() != exec) {
+        vec = experimental::distributed::Vector<ValueType>::create(
+            exec, comm, global_size, local_size);
+    } else if (vec->get_local_vector()->get_size() != local_size) {
+        // handle locally to eliminate the mpi call
+        std::cout << "local diff" << comm.rank() << std::endl;
```
Comment (pratikvn) [inline core/base/vector_cache.cpp:24]:
```suggestion
```

File: core/base/vector_cache.cpp [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        vec->local_ =
+            std::move(gko::matrix::Dense<ValueType>(exec, local_size));
```
Comment (pratikvn) [inline core/base/vector_cache.cpp:26]: nit: use the create function instead ?

File: core/base/vector_cache.cpp [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+{
+    if (!vec || vec->get_size() != template_vec->get_size() ||
+        vec->get_executor() != template_vec->get_executor()) {
+        vec =
+            experimental::distributed::Vector<ValueType>::create_with_config_of(
+                template_vec);
+    } else if (vec->get_local_vector()->get_size() !=
+               template_vec->get_local_vector()->get_size()) {
+        // handle locally to eliminate the mpi call
+        vec->local_ = std::move(gko::matrix::Dense<ValueType>(
```
Comment (pratikvn) [inline core/base/vector_cache.cpp:43]: nit: use the create function instead ?

File: include/ginkgo/core/matrix/dense.hpp [hunk 6]
```diff
@@ -122,6 +132,7 @@ class Dense
     friend class SparsityCsr<ValueType, int64>;
     friend class Dense<to_complex<ValueType>>;
     friend class experimental::distributed::Vector<ValueType>;
+    friend class gko::detail::VectorCache<ValueType>;
```
Comment (MarcelKoch) [inline include/ginkgo/core/matrix/dense.hpp:135]: why is this necessary?

**Review by MarcelKoch — APPROVED** [review]

**Review by pratikvn — APPROVED** [review]
Nice work! LGTM!

File: include/ginkgo/core/base/vector_cache.hpp [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+#include <ginkgo/config.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/mpi.hpp>
+#include <ginkgo/core/distributed/vector.hpp>
+
+
+#if GINKGO_BUILD_MPI
+
+
+namespace gko {
```
Comment (pratikvn) [inline include/ginkgo/core/base/vector_cache.hpp:20]: I think it might make sense to put this into the `experimental` and `distributed` namespaces.

File: include/ginkgo/core/base/vector_cache.hpp [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        return vec.get();
+    }
+};
+
+
+#endif
+
+
+}  // namespace detail
+}  // namespace gko
```
Comment (pratikvn) [inline include/ginkgo/core/base/vector_cache.hpp:111]:
```suggestion
}  // namespace detail
}  // namespace gko


#endif
```

**Review by pratikvn — CHANGES_REQUESTED** [review]
Some fixes needed

File: include/ginkgo/core/matrix/dense.hpp (hunk 6 above)
Comment (yhmtsai) [reply include/ginkgo/core/matrix/dense.hpp:135]: Because `distributed::Vector` holds the `Dense` object not `Dense` pointer, I need the constructor to construct the object directly and then use the move_to

File: core/base/vector_cache.cpp (hunk 4 above)
Comment (yhmtsai) [reply core/base/vector_cache.cpp:26]: We hold the object not the pointer.
However, I can create the unique pointer and release the object out actually, right?

File: include/ginkgo/core/matrix/dense.hpp (hunk 6 above)
Comment (yhmtsai) [reply include/ginkgo/core/matrix/dense.hpp:135]: ah, I might use the `unique_ptr.release` from create function

Comment (MarcelKoch) [reply include/ginkgo/core/matrix/dense.hpp:135]: No it's fine, I just got confused. I thought this was in the `distributed::Matrix`.

File: include/ginkgo/core/base/vector_cache.hpp (hunk 7 above)
Comment (yhmtsai) [reply include/ginkgo/core/base/vector_cache.hpp]: I move it into `gko::experimental::distributed::detail`

File: core/base/vector_cache.cpp (hunk 4 above)
Comment (pratikvn) [reply core/base/vector_cache.cpp]: No, It is a only a nit from my side. Using the constructor should also be okay.

**Review by pratikvn — APPROVED** [review]

File: include/ginkgo/core/distributed/preconditioner/schwarz.hpp [hunk 9]
```diff
@@ -150,6 +161,8 @@ class Schwarz
     void set_solver(std::shared_ptr<const LinOp> new_solver);
 
     std::shared_ptr<const LinOp> local_solver_;
+
+    gko::experimental::distributed::detail::VectorCache<ValueType> cache_;
```
[the remark this answers is not in the document]
Comment (yhmtsai) [reply include/ginkgo/core/distributed/preconditioner/schwarz.hpp]: it works. I did not test it before uploading, so I pick the one definitely works.
However, we need to use `gko::detail::...` not `detail::...` to call the function when in distributed namespace.
compiler only searches `distributed::detail` not `gko::detail`
